test/test_signal.py

```diff
@@ -0,0 +1,277 @@
+    def test_function():
+        return 1
+
+    object_id = test_function.remote()
+    ray.get(object_id)
+    result_list = signal.receive([signal.task_id(object_id)], timeout=5)
+    assert len(result_list) == 1
+    assert type(result_list[0][1]) == signal.DoneSignal
+
+
+def test_task_crash(ray_start):
+    # Get an error when ray.get() is called on the return of a failed task.
+
+    @ray.remote
+    def crashing_function():
+        raise Exception("exception message")
+
+    object_id = crashing_function.remote()
+    try:
+        ray.get(object_id)
+    except Exception as e:
+        assert type(e) == ray.worker.RayTaskError
+    finally:
+        result_list = signal.receive([(object_id)], timeout=5)
+        assert len(result_list) == 1
+        assert type(result_list[0][1]) == signal.ErrorSignal
+
+
+def test_task_crash_without_get(ray_start):
+    # Get an error when task failed.
+
+    @ray.remote
+    def crashing_function():
+        raise Exception("exception message")
+
+    object_id = crashing_function.remote()
+    result_list = signal.receive([(object_id)], timeout=5)
+    assert len(result_list) == 1
+    assert type(result_list[0][1]) == signal.ErrorSignal
+
+
+def test_actor_crash(ray_start):
+    # Get an error when ray.get() is called on a return parameter
+    # of a method that filed.
+
+    @ray.remote
+    class Actor(object):
+        def __init__(self):
+            pass
+        def crash(self):
+            raise Exception("exception message")
+
+    a = Actor.remote()
+    try:
+        ray.get(a.crash.remote())
+    except Exception as e:
+        assert type(e) == ray.worker.RayTaskError
+    finally:
+        result_list = signal.receive([a], timeout=5)
+        assert len(result_list) == 1
+        assert type(result_list[0][1]) == signal.ErrorSignal
+
+
+def test_actor_crash_init(ray_start):
+    # Get an error when an actor's __init__ failed.
+
+    @ray.remote
+    class ActorCrashInit(object):
+        def __init__(self):
+            raise Exception("exception message")
+        def m(self):
+            return 1
+
+    # Using try and catch won't catch the exception in the __init__.
+    a = ActorCrashInit.remote()
+    time.sleep(2)
+    result_list = signal.receive([a], timeout = 5)
+    assert len(result_list) == 1
+    assert type(result_list[0][1]) == signal.ErrorSignal
+
+
+def test_actor_crash_init2(ray_start):
+    # Get errors when (1) __init__ fails, and (2) subsequently when ray.get() is
+    # called on the return parameter of another method of the actor.
+
+    @ray.remote
+    class ActorCrashInit(object):
+        def __init__(self):
+            raise Exception("exception message")
+        def method(self):
+            return 1
+
+    a = ActorCrashInit.remote()
+    try:
+        ray.get(a.method.remote())
+    except Exception as e:
+        assert type(e) == ray.worker.RayTaskError
+    finally:
+        # TODO: this needs to be removed
+        time.sleep(5)
```

Comment:
This needs to be removed (at the moment the test doesn't seem to pass if I do), it indicates there might be some race condition